### DISCUSSION TITLE
Updating image.tag in helm values.yaml on gitlab, where image.tag refers to private gitlab container registry

### DISCUSSION THREAD

**Discussion by btxbtx** [renovatebot/renovate#17093] [Request Help] [answered]
Newbie to both Renovate and JSON.  I've been reading the documentation and some discussions here but am still stuck.  I will describe the intended use case below, followed by the code I have so far.

**Use Case**
I have three private repos on GitLab, whose `tree`s will be below:
- `repoA` which contain helm charts, as well as some `overrides` to override default chart values
- `repoB` which contains renovate config and the gitlab runner
- `repoC` which contains a private container registry with some docker images in it

I want `repoB` to be compare `repoA`'s `overrides/<chart name>/values.yaml` files on branch `develop` with the image tags of `repoC`'s images.  MRs should be opened that update `image.tag`s on `repoA` within each of those override values files.  I want these updates to happen every 15 minutes.

`image.tag` in `repoA` refers to the tag of an image from `repoC`.  The full image name is `repoC:<semver>`

**`repoA`**
```
├── README.md
├── charts
│   ├── chart1
│   │   ├── Chart.yaml
│   │   ├── templates
│   │   └── values.yaml
│   ├── chart2
│   │   ├── Chart.yaml
│   │   ├── templates
│   │   └── values.yaml
├── overrides
│   ├── chart1
│   │   ├── helmfile1
│   │   └── values.yaml
│   ├── chart2
│   │   ├── helmfile2
│   │   └── values.yaml
└── .gitlab-ci.yml
```

**`repoB`**
```
├── config.js
├── renovate.json
├── .gitlab-ci.yml
```

`repoC`'s tree intentionally omitted.

Here is what I have for `repoB`'s `config.js`
```js
module.exports = {
    token: "valid-gitlab-token",
    onboardingConfig: {
      extends: ["config:base"],
    },
    platform: "gitlab",

    repositories: [
        "path/to/repoA",
      ],
    
    baseBranches: ["develop"],

    "helm-values": {
        "fileMatch": ["overrides/*values.yaml"]
    },

    packageRules: [
      {
        matchManagers: ["helm-values"],
      }
    ],

    "schedule": [
        "every 15 minutes"
    ],
  };
  ```
  
`renovate.json`
```json
{
    "$schema": "https://docs.renovatebot.com/renovate-schema.json",
    "extends": [
      "config:base"
    ]
  }
  ```
  
To test the above config, I am running renovate locally:

```
$ renovate --dry-run
 WARN: cli config dryRun property has been changed to full
 INFO: Repository started (repository=path/to/repoA)
       "renovateVersion": "32.153.0"
 INFO: DRY-RUN: Would commit files to onboarding branch (repository=path/to/repoA)
 INFO: Dependency extraction complete (repository=path/to/repoA)
       "baseBranch": "develop",
       "stats": {
         "managers": {
           "gitlabci": {"fileCount": 1, "depCount": 1},
           "gitlabci-include": {"fileCount": 1, "depCount": 1},
           "helm-values": {"fileCount": 13, "depCount": 13},
           "helmfile": {"fileCount": 2, "depCount": 4},
           "helmv3": {"fileCount": 24, "depCount": 24}
         },
         "total": {"fileCount": 41, "depCount": 43}
WARN: No github.com token has been configured. Skipping release notes retrieval (repository=path/to/repoA)
       "manager": "helmfile",
       "depName": "sealed-secrets",
       "sourceUrl": "https://github.com/bitnami-labs/sealed-secrets"
 INFO: DRY-RUN: Would check branch renovate/configure (repository=path/to/repoA)
 INFO: DRY-RUN: Would create onboarding PR (repository=path/to/repoA)
 INFO: DRY-RUN: Would close Dependency Dashboard (repository=path/to/repoA)
       "title": "Dependency Dashboard"
 INFO: Repository finished (repository=path/to/repoA)
       "durationMs": 193665
  ```
  
You can probably tell that I am confused by a few things.  I would appreciate being pointed in the right direction to get my configuration working.

**Comment by viceice** [comment, marked as answer]
1. remove package rule, it does nothing, but issues a warning because it's invalid.
2. remove renovate stage from pipeline, it overrides default
3. provide debug logs, it't not clear whats going wrong 🤷‍♂️

**Reply by btxbtx** [reply]
Thank you for the feedback @viceice -- I have updated the post to try and reduce scope while giving more information.  Worth noting is I removed the gitlab runner aspect from this post, to instead focus on config troubleshooting.  I additionally added debug output when running renovate locally.  Please let me know if there's anything else you need from me and I will happily provide the details.

You mentioned I should remove the package rule because it is invalid and would issue a warning.  I do not see the warning in my debug output, however I do notice that a "helmfile" manager is being output.  I am hoping to update the `image.tag` inside of the `override/*values.yaml` files, and was trying to setup the `packageRules` to implement that behavior.

**Reply by viceice** [reply]
1. still no DEBUG logs, run `LOG_LEVEL=debug renovate ...`
2. post a trimed sample `values.yaml` file
3. the package rule is still wrong, what is your intension it should do?

**Reply by btxbtx** [reply]
Thank you for the prompt responses viceice.

I have tried to trim down the log file for you, it was quite large.

https://gist.github.com/btxbtx/67e19744b48b4aa8397e2f0920867745

A trimmed values file looks like:

```
image:
  registry: "registry.gitlab.com"
  repository: "path/to/repoA"
  tag: "latest"
  pullPolicy: "Always"
  pullSecrets:
    - "secret here"

env:
  KEY: value
```

And, given the following tree,

```
├── README.md
├── charts
│   ├── chart1
│   │   ├── Chart.yaml
│   │   ├── templates
│   │   └── values.yaml
│   ├── chart2
│   │   ├── Chart.yaml
│   │   ├── templates
│   │   └── values.yaml
├── overrides
│   ├── chart1
│   │   ├── helmfile1
│   │   └── values.yaml
│   ├── chart2
│   │   ├── helmfile2
│   │   └── values.yaml
└── .gitlab-ci.yml
```

My intention with the package rule/manager is that Renovate only tries to update `image.tag` inside of `overrides/<chart>/values.yaml` -- nothing else, not even `charts/<chart>/values.yaml`.

Let me know if something is unclear and I will try to explain better.

**Reply by viceice** [reply]
Renovate can't update `latest` docker tag. what do you expect to happen?

```
         "helm-values": [
           {
             "packageFile": "charts/<some project>/values.yaml",
             "deps": [
               {
                 "depName": "registry.gitlab.com/projectA/<some project>/projectA-<some project>",
                 "currentValue": "latest",
                 "datasource": "docker",
                 "replaceString": "latest",
                 "versioning": "docker",
                 "autoReplaceStringTemplate": "{{newValue}}{{#if newDigest}}@{{newDigest}}{{/if}}",
                 "depIndex": 0,
                 "updates": [],
                 "warnings": [],
                 "skipReason": "invalid-value"
               }
             ]
           },
```
☝️

**Reply by btxbtx** [reply]
I realize now that my sample values file was a bad example.  Some of our values files use `latest`, but some use a semver tag.  I am hoping that renovate can update those tags which use semver

**Reply by btxbtx** [reply]
I've noticed in the output pasted above, the `packageFile` is `charts/<some project>/values.yaml` -- part of what I am trying to configure is that Renovate ignore the `charts/` dir and instead only look at `overrides/`

**Reply by btxbtx** [reply]
@viceice I have made some progress here, but am still stuck on something.  Would it be proper to open a new discussion, or continue on here?

```js
module.exports = {
    token: "xxxxxxxxx",
    onboardingConfig: {
      globalExtends: ["config:base"],
    },
    platform: "gitlab",

    repositories: [
        "path/to/repoA",
      ],
    
    baseBranches: ["develop"],
    
    labels: ["dependencies", "renovate"],

    "helm-values": {
        "fileMatch": ["overrides/*values.yaml"],
    },

    "ignorePaths": ["**charts/**"],

    "enabledManagers": ["helm-values"],

    "schedule": [
        "every 15 minutes"
    ],
  };
```

I also have this `renovate.json` in `repoA`:

```json
{
    "$schema": "https://docs.renovatebot.com/renovate-schema.json",
    "extends": [
      "config:base"
    ]
  }
  ```

However, the combination of `"helm-values"` and `ignorePaths` in `config.js` is not working as expected.  I am expecting to only match values.yaml files inside the `overrides` dir, not any inside of `charts` dirs.

It looks like `ignorePaths` is being overriden by the `renovate.json` in repoA.  Some relevant logs below.


`ignorePaths` being overriden:
```
DEBUG: File config
       "config": {
         "token": "***********",
         "onboardingConfig": {"globalExtends": ["config:base"]},
         "platform": "gitlab",
         "repositories": ["path/to/repoA"],
         "baseBranches": ["develop"],
         "labels": ["dependencies", "renovate"],
         "helm-values": {"fileMatch": ["overrides/*values.yaml"]},
         "ignorePaths": ["**charts/**"],
         "enabledManagers": ["helm-values"],
         "schedule": ["every 15 minutes"]
       }
       DEBUG: Combined config
       "config": {
         "token": "***********",
         "onboardingConfig": {"globalExtends": ["config:base"]},
         "platform": "gitlab",
         "repositories": ["path/to/repoA"],
         "baseBranches": ["develop"],
         "labels": ["dependencies", "renovate"],
         "helm-values": {"fileMatch": ["overrides/*values.yaml"]},
         "ignorePaths": ["**charts/**"],
         "enabledManagers": ["helm-values"],
         "schedule": ["every 15 minutes"],
         "hostRules": [],
         "dryRun": "full"
       }
       
      DEBUG: Found repo ignorePaths (repository=path/to/repoA)
       "ignorePaths": [
         "**/node_modules/**",
         "**/bower_components/**",
         "**/vendor/**",
         "**/examples/**",
         "**/__tests__/**",
         "**/test/**",
         "**/tests/**",
         "**/__fixtures__/**"
       ]
  ```

`overrides/` dir not being matched:
```
DEBUG: Using file match: (^|/)values\.yaml$ for manager helm-values (repository=path/to/repoA)
DEBUG: Using file match: overrides/*values.yaml for manager helm-values (repository=path/to/repoA)
DEBUG: Matched 24 file(s) for manager helm-values: charts/<some project>/values.yaml, charts/<some other project>.yaml, <more charts/ values, no overrides/ values> (repository=path/to/repoA)
DEBUG: Found helm-values package files (repository=path/to/repoA)
DEBUG: Found 13 package file(s) (repository=path/to/repoA)
 INFO: Dependency extraction complete (repository=path/to/repoA)
       "baseBranch": "develop",
       "stats": {
         "managers": {"helm-values": {"fileCount": 13, "depCount": 13}},
         "total": {"fileCount": 13, "depCount": 13}
       }
       
       DEBUG: packageFiles with updates (repository=path/to/repoA)
       "config": {
         "helm-values": [
           {
             "packageFile": "charts/<some project>/values.yaml",
             "deps": [
               {
                 "depName": "registry.gitlab.com/tacen/<some project>",
                 "currentValue": "latest",
                 "datasource": "docker",
                 "replaceString": "latest",
                 "versioning": "docker",
                 "autoReplaceStringTemplate": "{{newValue}}{{#if newDigest}}@{{newDigest}}{{/if}}",
                 "depIndex": 0,
                 "updates": [],
                 "warnings": [],
                 "skipReason": "invalid-value"
               }
             ]
           },
```

Ignoring the `invalid-value`, `charts/` is supposed to be ignored and `overrides` is supposed to be searched for matches.

**Reply by viceice** [reply]
```json
...
"fileMatch": ["^overrides\\/.+\\/values\\.yaml$"],
...
```

wrong `fileMatch` regex

**Reply by viceice** [reply]
updated comment above

**Reply by viceice** [reply]
remove `ignorePaths`

**Reply by btxbtx** [reply]
@viceice thank you for the suggestion, it looks like both fileMatches are being used still, but `overrides` is not actually matching:

```
DEBUG: Using file match: (^|/)values\.yaml$ for manager helm-values (repository=path/to/repoA)
DEBUG: Using file match: ^overrides\/.+\/values\.yaml$ for manager helm-values (repository=path/to/repoA)
DEBUG: Matched 24 file(s) for manager helm-values: charts/<some project>/values.yaml, charts/<some other project>.yaml, <more charts/ values, no overrides/ values> (repository=path/to/repoA)
DEBUG: Found helm-values package files (repository=path/to/repoA)
```

```js
module.exports = {
    token: "xxxxxxxx",
    onboardingConfig: {
      globalExtends: ["config:base"],
    },
    platform: "gitlab",

    repositories: [
        "path/to/repoA",
      ],
    
    baseBranches: ["develop"],
    
    labels: ["dependencies", "renovate"],

    "helm-values": {
      "fileMatch": ["^overrides\\/.+\\/values\\.yaml$"],
    },

    "enabledManagers": ["helm-values"],

    "schedule": [
        "every 15 minutes"
    ],
  };
  ```

**Reply by btxbtx** [reply]
moving the `ignorePaths` to `repoA` solved the issue.  Thank you for all the help!

**Reply by viceice** [reply]
yes `fileMatch`[^1] is mergable, so it will merge both values

[^1]: https://docs.renovatebot.com/configuration-options/#filematch

**Reply by viceice** [reply]
you can use `matchFiles`[^1] and `matchPaths`[^2] in packages rules to limit to specific managers, `ignorePaths` will disable all managers on that paths


[^1]: https://docs.renovatebot.com/configuration-options/#matchfiles
[^2]: https://docs.renovatebot.com/configuration-options/#matchpaths
